File: src/main/java/editor/gui/generic/OverwriteFileChooser.java
package editor.gui.generic;

import java.io.File;
import java.util.Arrays;

import javax.swing.JFileChooser;
import javax.swing.JOptionPane;
import javax.swing.filechooser.FileNameExtensionFilter;

/**
 * This is a {@link JFileChooser} that confirms first if a file should be overwritten.
 *
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class OverwriteFileChooser extends JFileChooser
{
    /**
     * Create a new OverwriteFileChooser.
     */
    public OverwriteFileChooser()
    {
        super();
    }

    /**
     * Create a new OverwriteFileChooser that starts in the given path
     *
     * @param currentDirectoryPath path to start in
     */
    public OverwriteFileChooser(String currentDirectoryPath)
    {
        super(currentDirectoryPath);
    }

    /**
     * {@inheritDoc}
     * <p>
     * If the file name doesn't have an extension specified by the file filter, add it.
     */
    @Override
    public File getSelectedFile()
    {
        if (getFileFilter() instanceof FileNameExtensionFilter)
        {
            FileNameExtensionFilter filter = (FileNameExtensionFilter)getFileFilter();
            String fname = super.getSelectedFile().getAbsolutePath();
            if (Arrays.stream(filter.getExtensions()).noneMatch((ext) -> fname.endsWith("." + ext)))
                return new File(fname + "." + filter.getExtensions()[0]);
        }
        return super.getSelectedFile();
    }

    /**
     * {@inheritDoc}
     * <p>
     * If the file already exists, make sure the user wants to replace it first. If not,
     * allow them to make another choice or close the dialog depending on what was
     * selected.
     */
    @Override
    public void approveSelection()
    {
        if (getSelectedFile().exists() && getDialogType() == SAVE_DIALOG)
        {
            switch (JOptionPane.showConfirmDialog(this, "File already exists.  Overwrite?", "Warning", JOptionPane.YES_NO_CANCEL_OPTION))
            {
            case JOptionPane.YES_OPTION:
                super.approveSelection();
                return;
            case JOptionPane.NO_OPTION:
                return;
            case JOptionPane.CANCEL_OPTION:
                cancelSelection();
                return;
            case JOptionPane.CLOSED_OPTION:
                return;
            }
        }
        super.approveSelection();
    }
}

File: src/main/scala/editor/gui/generic/OverwriteFileChooser.scala
package editor.gui.generic

import java.io.File
import javax.swing.JFileChooser
import javax.swing.JOptionPane
import javax.swing.filechooser.FileNameExtensionFilter

/**
 * File chooser that asks permission to replace existing files. A "Yes" choice replaces the file,
 * "No" returns to the chooser for renaming, and "Cancel" or quit simply exits without doing any
 * file operations.
 * 
 * @constructor create a new file chooser that asks permission to overwrite files
 * @param cwd starting directory of the file chooser, or user's default directory (as specified by
 * [[JFileChooser]]) if not set
 * 
 * @author Alec Roelke
 */
class OverwriteFileChooser(cwd: String = JFileChooser().getCurrentDirectory.getAbsolutePath) extends JFileChooser(cwd) {
  override def getSelectedFile = {
    Option(super.getSelectedFile).map{ file => getFileFilter match {
      case filter: FileNameExtensionFilter if !filter.getExtensions.exists((ext) => file.getAbsolutePath.endsWith(s".$ext")) => File(s"${file.getAbsolutePath}.${filter.getExtensions.apply(0)}")
      case _ => file
    }}.getOrElse(null)
  }

  override def approveSelection() = {
    if (getSelectedFile.exists && getDialogType == JFileChooser.SAVE_DIALOG) {
      JOptionPane.showConfirmDialog(this, "File already exists. Overwrite?", "Warning", JOptionPane.YES_NO_CANCEL_OPTION) match {
        case JOptionPane.YES_OPTION => super.approveSelection
        case JOptionPane.NO_OPTION =>
        case JOptionPane.CANCEL_OPTION => cancelSelection()
        case JOptionPane.CLOSED_OPTION =>
      }
    } else {
      super.approveSelection()
    }
  }
}